### PR TITLE
Billing service implementation

### PR DESCRIPTION
**Pull request by ivanr3**
This PR adds an implementation of Billing Service. The PR is for corda-v4 branch (as Billing Service requires Reference States). 
I'll merge corda-v4 to the master when C4 is officially released.

### CODE REVIEW

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/Constants.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.r3.businessnetworks.billing.flows
+
+import net.corda.core.utilities.seconds
+
+object Constants {
+    /**
+     * Default tim tolerance that is used for transaction time windows
```
Comment (davejh69) [inline bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/Constants.kt]: s/tim/time/c?

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt [hunk 2]
```diff
@@ -13,23 +14,22 @@ import net.corda.core.flows.InitiatingFlow
...
-import net.corda.core.utilities.seconds
 import java.time.Duration
 import java.time.Instant
 
 /**
  * Chips off one or more states from the BillingState
  *
  * @param billingState state to chip off from
- * @param amount amount of each cheap off state
+ * @param chipAmount amount of each cheap off state
```
Comment (davejh69) [inline bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt]: cheap?

File: bn-apps/billing/billing-app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,61 @@
+apply plugin: 'kotlin'
+apply plugin: 'net.corda.plugins.cordapp'
+apply plugin: 'net.corda.plugins.cordformation'
+apply plugin: 'net.corda.plugins.quasar-utils'
+
+cordapp {
+    targetPlatformVersion 3
+    minimumPlatformVersion 1
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-app/build.gradle]: Hm... For some reason sometimes I get comments on outdated changes. For the billing app, the min  is actually 4, as the contract expect reference states. How come the changeset showed an outdated version I wonder... I've changed that a while ago: https://github.com/corda/corda-solutions/blob/billing-service-implementation/bn-apps/billing/billing-app/build.gradle#L7

File: bn-apps/billing/billing-app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Corda integration dependencies
+    cordaCompile "$corda_release_group:corda-core:$corda_release_version"
+    cordaCompile "$corda_release_group:corda-rpc:$corda_release_version"
+    cordaCompile "$corda_release_group:corda-node-api:$corda_release_version"
+    cordaRuntime "$corda_release_group:corda:$corda_release_version"
+    testCompile "$corda_release_group:corda-node-driver:$corda_release_version"
+
+    cordapp project(":bn-apps:memberships-management:membership-service-contracts-and-states")
+
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-app/build.gradle]: removed :)

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/bno/IssueBillingStateFlow.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import net.corda.core.identity.Party
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+
+@InitiatingFlow
+class IssueBillingStateFlow(private val owner : Party,
+                            private val amount : Long) : FlowLogic<Pair<BillingState, SignedTransaction>>() {
+    @Suspendable
+    override fun call() : Pair<BillingState, SignedTransaction> {
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/bno/IssueBillingStateFlow.kt]: Good spot. FOr some reason it showed an outdated changes again: https://github.com/corda/corda-solutions/blob/billing-service-implementation/bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/bno/IssueBillingStateFlow.kt#L38

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/AttachUnspentChipsFlowResponder.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            builder.addOutputState(billingState.state.data.copy(spent = billingState.state.data.spent - totalUnspent), BillingContract.CONTRACT_NAME)
+
+            builder.verify(serviceHub)
+
+            val stx = serviceHub.signInitialTransaction(builder)
+
+            val allSignedTx = subFlow(CollectSignaturesFlow(stx, listOf(session)))
+            val finalisedTx = subFlow(FinalityFlow(allSignedTx, listOf(session)))
+
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/AttachUnspentChipsFlowResponder.kt]: Hah, that's the default Intellij-Kotlin formattign style :)

File: bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt [hunk 7]
```diff
@@ -1,104 +1,159 @@
...
-                }
+        val commands = tx.commandsOfType<Commands>()
+
+        if (commands.isEmpty()) {
+            throw IllegalArgumentException("Transaction must contain at least one BillingContract command")
+        }
+
+        if (commands.first().value is Commands.SpendChip) {
+            verifySpendTransaction(tx, commands)
+        } else {
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt]: Hm good point actually. Fixed.

File: bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt [hunk 8]
```diff
@@ -1,104 +1,159 @@
...
+
+        "Issued amounts of billing states should be equal" using (inputBillingState.issued == outputBillingState.issued)
+        "Issuer of billing states should be equal" using (inputBillingState.issuer == outputBillingState.issuer)
+        "Linear id of billing states should be equal" using (inputBillingState.linearId == outputBillingState.linearId)
+        "Expiry date of billing states should be equal" using (inputBillingState.expiryDate == outputBillingState.expiryDate)
+        "Owner of billing states should be equal" using (inputBillingState.owner == outputBillingState.owner)
+        "Spent amount of the output billing state should be positive" using (outputBillingState.spent >= 0L)
+        "ChipOff transaction should be signed only by the owner" using (command.signers.single() == outputBillingState.owner.owningKey)
+
+
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt]: Fixed

File: bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt [hunk 9]
```diff
@@ -1,104 +1,159 @@
...
+        var totalAttach = 0L
+        chips.forEach {
+            "Owner of BillingChipState should be the same as of BillingStates" using (it.owner == outputBillingState.owner)
+            "Linear id of BillingChipState should be the same as of BillingState" using (it.billingStateLinearId == outputBillingState.linearId)
+            totalAttach += it.amount
+        }
+
+        "Attached amounts should match" using (inputBillingState.spent - totalAttach == outputBillingState.spent)
+    }
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt]: Thanks!

File: bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt [hunk 10]
```diff
@@ -13,115 +13,124 @@ class BillingContract : Contract {
...
+            "Spent amount of the output BillingState should be less or equal to the issued" using (outputBillingState.spent <= outputBillingState.issued)
         }
         if (outputBillingState.expiryDate != null) {
             val timeWindow = tx.timeWindow!!
-            "Billing state expiry date should be within the specified time window" using (timeWindow.contains(outputBillingState.expiryDate))
+            "Output BillingState expiry date should be within the specified time window" using (timeWindow.contains(outputBillingState.expiryDate))
+        }
+    }
+
+
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-contracts-and-states/src/main/kotlin/com/r3/businessnetworks/billing/states/BillingState.kt]: Removed

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt [hunk 11]
```diff
@@ -34,6 +38,11 @@ class ChipOffBillingStateFlow(private val billingState : StateAndRef<BillingStat
                 .addInputState(billingState)
                 .addCommand(BillingContract.Commands.ChipOff(), billingState.state.data.owner.owningKey)
 
+        // if expiry date is present - adding a time window
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt]: fixed

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/Constants.kt (hunk 1 above)
Comment (ivanr3) [reply bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/Constants.kt]: fixed

File: bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt (hunk 2 above)
Comment (ivanr3) [reply bn-apps/billing/billing-app/src/main/kotlin/com/r3/businessnetworks/billing/flows/member/ChipOffBillingStateFlow.kt]: fixed

**Pull request closed without merging**